docs/design.md

```diff
@@ -0,0 +1,785 @@
+* `replace` - overwrite data at the specified path and replace it with the data
+  given in this document
+* `delete` - remove the data at the specified path
+
+After actions are applied for a given layer, substitutions are applied (see
+the Substitution section for details).
+
+Selection of document parents is controlled by the `parentSelector` field and
+works as follows. A given document, `C`, that specifies a `parentSelector`
+will have exactly one parent, `P`. Document `P` will be the highest
+precedence (i.e. part of the lowest layer) document that has the labels
+indicated by the `parentSelector` (and possibly additional labels) from the
+set of all documents of the same `schema` as `C` that are in layers above the
+layer `C` is in. For example, consider the following sample documents:
+
+```yaml
+---
+schema: deckhand/LayeringPolicy/v1
+metadata:
+  schema: metadata/Control/v1
+  name: layering-policy
+data:
+  layerOrder:
+    - global
+    - region
+    - site
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: global-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: global
+data:
+  a:
+    x: 1
+    y: 2
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: region-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: region
+    parentSelector:
+      key1: value1
+    actions:
+      - method: replace
+        path: .a
+data:
+  a:
+    z: 3
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: site-1234
+  layeringDefinition:
+    layer: site
+    parentSelector:
+      key1: value1
+    actions:
+      - method: merge
+        path: .
+data:
+  b: 4
+...
+```
+
+When rendering, the parent chosen for `site-1234` will be `region-1234`,
+since it is the highest precedence document that matches the label selector
+defined by `parentSelector`, and the parent chosen for `region-1234` will be
+`global-1234` for the same reason. The rendered result for `site-1234` would
+be:
+
+```yaml
+---
+schema: example/Kind/v1
+metadata:
+  name: site-1234
+data:
+  a:
+    z: 3
+  b: 4
+...
+```
+
+If `region-1234` were later removed, then the parent chosen for `site-1234`
+would become `global-1234`, and the rendered result would become:
+
+```yaml
+---
+schema: example/Kind/v1
+metadata:
+  name: site-1234
+data:
+  a:
+    x: 1
+    y: 2
+  b: 4
+...
+```
+
+<!-- TODO: Add figures for this example, with region present, have site point
+with dotted line at global and indicate in caption (or something) that it's
+selected for but ignored, because there's a higher-precedence layer to select
+-->
+
+### Substitution
+
+Substitution is primarily designed as a mechanism for inserting secrets into
+configuration documents, but works for unencrypted source documents as well.
+Substitution is applied at each layer after all merge actions occur.
```

Comment:
If during the merge phase sections of the finalized document whose data section has been changed, then this could mean that previous references to sections in the data section by lower-precedence documents are no longer valid, i.e. point to non-existent sections. Is this something that is ignored?